### ISSUE TITLE
docs: httpBatchLink usage example comes across ambiguous

### ISSUE THREAD

**Issue by henryandrewm** [trpc/trpc#3847]
### Area of Improvement

The usage example makes it seem like only multiple requests for the same sub router and procedure will be batched together. This is not the case in Next.js though.

### Link to related docs

https://trpc.io/docs/links/httpBatchLink

### Additional information

The usage example shows requests from the same sub router procedure with different query inputs which is fine, but it probably needs to be mentioned that all requests from a functional component, regardless of if they are for the same sub router and procedure, will be batched together. This may actually be a bug, but it's happening in this package version combination

@tanstack/react-query: ^4.24.9 => 4.24.9 
@trpc/client: ^10.12.0 => 10.12.0 
@trpc/next: ^10.12.0 => 10.12.0 
@trpc/react-query: ^10.12.0 => 10.12.0 
@trpc/server: ^10.12.0 => 10.12.0 
next: ^13.1.1 => 13.1.1 
react: ^18.2.0 => 18.2.0 
typescript: ^4.9.4 => 4.9.4

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR implementing the suggested changes!

**Comment by iduuck** [comment]
Is your claim, that the following snippet suggests the issue for you? 

```
const somePosts = await Promise.all([
  trpc.post.byId.query(1);
  trpc.post.byId.query(2);
  trpc.post.byId.query(3);
])
```

**Comment by henryandrewm** [comment]
Correct. Along with that the snippet and description before it suggests that the only way for the requests to be batched are to put them in a `Promise.all`. If you just had several requests in a row from a functional component though it automatically batches them, even without the `Promise.all`. With this, it may only return one batched response from the server, but it could make more than one database transaction if you are calling different procedures.

**Comment by iduuck** [comment]
Yeah, you are right. Actually the code snippet that I pasted is dedicated for the server only (see the `.query` instead of `.useQuery`).

I will check if the documentation can be tweaked a little. However, on the server **afaik** there is no request being made on the server. The `Promise.all` is only hindering the server to wait for the first Promise to fill, until it goes into the next Promise.

**Comment by henryandrewm** [comment]
That's completely fair. I'm using `.useQuery` in my project so that was a bit of an oversight on my side.

Before any potential changes are made though, I did want to ask if `httpBatchLink` was meant to batch all requests together or just multiple requests to the same procedure? If it's the latter then I should probably report this as a bug in Next.js, correct?

**Comment by henryandrewm** [comment]
Thank you for the explanation, that definitely clears things up for me! I'm realizing now that the main description of `httpBatchLink` on that page was what created my confusion. It says 

`batches an array of individual tRPC operations into a single HTTP request that's sent to a single tRPC procedure.`

I think if this was tweaked to be written like your description in the last comment it would be more straightforward. The ending of that sentence makes it seem like it batches for single procedures only instead of all requests that come in during a given time frame.

Regardless, thank you for taking the time to look at this!

**Comment by KATT** [comment]
React-query automatically handles deduping. If you're not using react query, you can do your own `dedupeLink`, we even have one internally for testing but haven't exported it: https://github.com/trpc/trpc/blob/9d5561d1b6b74c7ac9df393f6a3344cf7d93b7f2/packages/client/src/links/dedupeLink.ts

**Comment by KATT** [comment]
@henryandrewm would you be keen on doing a contribution to the docs? :)

**Comment by slhck** [comment]
There's an additional part that gets me very confused in the docs:

> The code below will produce exactly one HTTP request and on the server exactly one database query:

How exactly is it guaranteed that the server only executes one database query here? What has to be done when defining procedures to actually make this work (i.e. getting mulitple requests at once, if I were to load, say, all the posts instead of one-by-one)? I am basically missing the server example in the docs.

---

In my particular case with a Next.js app, I have a procedure that I call with batching:

```ts
  const { data: automatorConfig } = trpc.getAutomatorConfigForLabel.useQuery(
    {
      label,
    },
    {
      trpc: {
        context: {
          useBatch: true,
        },
      },
    },
  );
```

This is part of a component that is renderered multiple times on the same page, for different labels, so this is indeed batched and creates a call like:

```
http://localhost:3008/api/trpc/getAutomatorConfigForLabel,getAutomatorConfigForLabel,getAutomatorConfigForLabel,getAutomatorConfigForLabel,getAutomatorConfigForLabel?batch=1&input=%7B%220%22%3A%7B%22label%22%3A%22client-001%22%7D%2C%221%22%3A%7B%22label%22%3A%22client-002%22%7D%2C%222%22%3A%7B%22label%22%3A%22client-003%22%7D%2C%223%22%3A%7B%22label%22%3A%22client-004%22%7D%2C%224%22%3A%7B%22label%22%3A%22client-005%22%7D%7D
```

This however leads to the procedure being called *multiple times*:

```ts
getAutomatorConfigForLabel: publicProcedure
  .input(
    z.object({
      label: z.string(),
    }),
  )
  .query(async ({ input }) => {
    const { label } = input;
    console.log('getAutomatorConfigForLabel', label);
    const config = database.getAutomatorConfigForLabel(label);
    return config;
  }),
```

the log shows:

```
getAutomatorConfigForLabel client-001
getAutomatorConfigForLabel client-002
getAutomatorConfigForLabel client-003
getAutomatorConfigForLabel client-004
```

So it's not much faster than disabling batching, because I still have to hit the database 4⨉ instead of just once with an array of `label`.

**Comment by simi-obs** [comment]
Yeah it is quite confusing. So @iduuck or @KATT , based on what I read here - I understand that the batching actually only saves time when it comes to creating "context", i.e when I call 
`initTRPC.context<ApiKeyProtectedRouterContext>().create()`

> This way, the context is only created one time, and you can do some sort of async function in the context, and this saves you some time.

How can execute async function as a part of context creation?

Because if I am using middlewares on the procedures which enrich the context, these I assume are called for each procedure in the batch separately. That means each "procedure" in a batch receives a copy of a context, correct?

**Comment by jel-massih** [comment]
> How exactly is it guaranteed that the server only executes one database query here? What has to be done when defining procedures to actually make this work (i.e. getting mulitple requests at once, if I were to load, say, all the posts instead of one-by-one)? I am basically missing the server example in the docs.

From what I understand, the "One Database Query" is not related to TRPC, but rather the assumption you are using Prisma (or some other orm that internally handles query aggregation). (KATT gave more info in https://github.com/trpc/trpc/discussions/921)

As we dont use Prisma, but wanted this type of behavior (effectively procedure aggregation where many single requests on client collapses to a single multiple request on server), we were able to accomplish this with a combination of creating a custom procedure remap link, and patching the HTTPBatchLink to internally aggregate queries. Ill give a quick overview of how we did this below.

Ex. You have `post.byId` which takes input {id: string} and `post.byIds` which takes {ids: string[]}

Following the docs you make a call like:
```
const somePosts = await Promise.all([
  trpc.post.byId.query({id: 1}),
  trpc.post.byId.query({id: 2}),
  trpc.post.byId.query({id: 3}),
]);
```
With HttpBatchLink it will make 3 procedure calls (and run that code for each id effectively)

We created a new "procedureRemapLink" that sits before httpBatchLink which will remap "single" procedures to the "multi" variant. (so if the operation is `{path: 'post.byId', input: {id: 1}}` it will modify the operation to `{path: 'post.byIds', input: {ids: [1]}}` On the way back out, it will restore the original operation, and extract the relevant result from the aggregated operations result array (since it will contain all 3 posts due to the aggregation in the following step).

Since with just the above remapping, it will still send 3 individual procedure calls to `post.byIds` with array of single id, we need to aggregate those 3 operations into a single operation with all the ids in the input. We do this by patching httpBatchLink dataLoader to before adding  an operation to the pending items, we check if we can aggregate it (where there is already an existing operation of this type, and instead of adding the operation to the pendingOps, we just add the input to the existing operation, and the ops resolve/reject to the existing items resolve/reject chain.

Coarsly things go like this:
procedureRemapLink Input:  `[{'post.byId': {id: 1}}, {'post.byId': {id: 2}}, {'post.byId': {id: 3}}]`
httpBatchLink Input: `[{'post.byIds': {ids: [1]}}, {'post.byIds': {ids: [2]}}, {'post.byIds': {ids: [3]}}]`
httpBatchLink HttpRequest: `[{'post.byIds': {ids: [1,2,3]}}]` (This is whats sent to server)
httpBatchLink Output: (same as input, but all 3 ops have the identical response data in format like `[{id: 1, content: 'Post 1'}, {id: 2, content: 'Post 2'}, {id: 3, content: 'Post 3'}]` 
procedureRemapLink output:  `[{'post.byId': {id: 1, content: 'Post 1'}}, {'post.byId': {id: 2, content: 'Post 2'}}, {'post.byId': {id: 3, content: 'Post 3'}}]` [Note that original input is restored, and just the relevant result is picked

Since the remapping is restored back to original on way out, the rest of the client doesn't need to know any of this happens and it `just works` of course lots of error and edge cases to handle, this is super simplified.


Apologies for super confusing and convoluted explanation. Hopefully this maybe can help someone trying to do something similar to what we are doing. If there is any interest I can try and spend some time to clean this up and make a publicly available Link that does the above.
